File: backend/src/utils/run-id.ts
/**
 * Shared correlation identity — runId (decision f1cbd5ef, architect design
 * "runId — Shared Correlation Identity Design", Pass 1).
 *
 * runId is SERVER-MINTED ONLY. This module is the sole TypeScript producer.
 * No code path may read a runId off an external/client request body — any
 * inbound client-supplied value must be stripped/ignored, mirroring how the
 * client-minted `orchestrationId` is already discarded in
 * `task-runner-resolver.ts` (submitTask mints its own via `randomUUID()`
 * and never reads one from `SubmitTaskInput`).
 *
 * Format: opaque string `run-<uuidv4>` (lowercase, hyphenated). The `run-`
 * prefix makes it greppable in logs/DDB and impossible to confuse with the
 * bare-uuid orchestrationId/executionId/projectId already in use.
 *
 * `DispatchContext` + `buildDispatchContext()` are the BUILD-TIME durability
 * guard: `runId` is a REQUIRED field on the type and a required parameter
 * on the builder, so a future entry point that omits it fails `tsc`
 * (compile-time), not merely a runtime check. Every Pass-1 entry point
 * routes its outbound envelope through this builder.
 */
import { randomUUID } from "crypto";

/** Sole producer: mints a fresh, server-side runId. Never derived from external input. */
export function mintRunId(): string {
  return `run-${randomUUID()}`;
}

/**
 * Shared outbound dispatch/event envelope. `runId` is REQUIRED — omitting
 * it at a call site is a compile-time (`tsc`) failure, which is the
 * build-time half of the silent-regression guard (design §3, layer 1).
 * Additional fields are intentionally open (`[key: string]: unknown`) so
 * this type can wrap the varied per-entry-point detail shapes without
 * forcing a single rigid schema.
 */
export interface DispatchContext {
  runId: string;
  [key: string]: unknown;
}

/**
 * Construct a `DispatchContext`. `runId` has no default — every caller must
 * supply one explicitly (typically via `mintRunId()`), so omission is a
 * compile error rather than a silently-undefined field.
 */
export function buildDispatchContext(
  context: DispatchContext,
): DispatchContext {
  return { ...context };
}

File: arbiter/common/run_id.py
"""Shared correlation identity — run_id (decision f1cbd5ef, architect design
"runId — Shared Correlation Identity Design", Pass 1).

run_id is SERVER-MINTED ONLY. This module is the sole Python producer,
mirroring ``backend/src/utils/run-id.ts`` on the TypeScript side. No code
path may read a run_id off an external/client request body — any inbound
client-supplied value must be stripped/ignored.

Format: opaque string ``run-<uuidv4>`` (lowercase, hyphenated) — identical
shape to the TS producer, so a run_id minted by either tier is
indistinguishable to a downstream consumer.

``build_dispatch_context()`` is the Python half of the BUILD-TIME durability
guard (design §3, layer 1): ``run_id`` is a REQUIRED keyword-only parameter
with no default, so a call site omitting it fails at call time with a
``TypeError`` (Python has no compile-time check, so this is enforced as
strictly as the language allows: an unconditional, un-catchable
``TypeError`` from the interpreter itself for a missing required argument,
not a caught/logged soft failure).
"""
from __future__ import annotations

import uuid
from typing import Any


def mint_run_id() -> str:
    """Sole producer: mint a fresh, server-side run_id. Never derived from
    external input."""
    return f"run-{uuid.uuid4()}"


def build_dispatch_context(*, run_id: str, **context: Any) -> dict:
    """Construct a dispatch/event envelope dict. ``run_id`` is a required
    keyword-only argument — omitting it raises ``TypeError`` immediately,
    which is the build-time-equivalent durability guard for the Python tier
    (mirrors the TS `DispatchContext` type's required `runId` field).
    """
    return {"run_id": run_id, **context}
